pkg/kubelet/kubelet.go

```diff
@@ -1452,48 +1460,70 @@ func (kl *Kubelet) SyncPods(allPods []*api.Pod, podSyncTypes map[types.UID]metri
 		return nil
 	}
 
-	// Kill any containers we don't need.
-	killed := []string{}
+	// Kill containers associated with unwanted pods in parallel.
+	type result struct {
+		containers []*docker.Container
+		err        error
+	}
+	ch := make(chan result, len(runningPods))
+	defer close(ch)
+	numWorkers := 0
 	for _, pod := range runningPods {
 		if _, found := desiredPods[pod.ID]; found {
-			// syncPod() will handle this one.
+			// Per-pod workers will handle the desired pods.
 			continue
 		}
-
-		// Kill all the containers in the unidentified pod.
-		for _, c := range pod.Containers {
-			// call the networking plugin for teardown
-			if c.Name == dockertools.PodInfraContainerName {
-				err := kl.networkPlugin.TearDownPod(pod.Namespace, pod.Name, dockertools.DockerID(c.ID))
-				if err != nil {
-					glog.Errorf("Network plugin pre-delete method returned an error: %v", err)
-				}
+		numWorkers++
+		go func(pod *kubecontainer.Pod, ch chan result) {
```

Comment:
I moved the whole block (including result aggregation) to a different function. This should improve the readability a little bit.